engine/runtime/src/main/java/org/enso/interpreter/runtime/system/System.java

```diff
@@ -52,8 +52,15 @@ public static long nanoTime() {
   @CompilerDirectives.TruffleBoundary
   public static void exit(long code) {
     var ctx = EnsoContext.get(null);
-    EnsoLanguage.get(null).disposeContext(ctx);
-    java.lang.System.exit((int) code);
+    if (ctx.isInteractiveMode()) {
+      // In interactive mode, the ExitException should be caught and handled by one of
+      // the instruments that should take care of proper context disposal.
+      throw new ExitException((int) code);
+    } else {
+      // While not in interactive mode, it is safe to directly call
+      // TruffleContext.exitContext
+      ctx.exit((int) code);
```

Comment:
Do we need this `ctx.exit`? Cannot we have just `ExitException`? What's the benefit of doing `ctx.exit`?

Reply:
Initially, There was only `ctx.exit` that just calls `TruffleContext.closeExited` which immediately starts context hard exit. However, in interactive mode, it caused some issues.

> Cannot we have just ExitException?

Probably yes. I will try to use only `ExitException` that implements the proper Truffle contract.